util/src/migration/tests.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2015, 2016 Ethcore (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::BTreeMap;
+use migration::{Error, Destination, Migration, Manager, Config};
+
+impl Destination for BTreeMap<Vec<u8>, Vec<u8>> {
+	fn commit(&mut self, batch: BTreeMap<Vec<u8>, Vec<u8>>) -> Result<(), Error> {
+		self.extend(batch);
+		Ok(())
+	}
+}
+
+struct Migration0;
+
+impl Migration for Migration0 {
+	fn from_version(&self) -> &'static str {
+		"0"
+	}
+
+	fn to_version(&self) -> &'static str {
+		"1"
+	}
+
+	fn simple_migrate(&self, key: Vec<u8>, value: Vec<u8>) -> Option<(Vec<u8>, Vec<u8>)> {
+		let mut key = key;
+		key.push(0x11);
+		let mut value = value;
+		value.push(0x22);
+		Some((key, value))
+	}
+}
+
+struct Migration1;
+
+impl Migration for Migration1 {
+	fn from_version(&self) -> &'static str {
+		"1"
+	}
+
+	fn to_version(&self) -> &'static str {
+		"3"
```

Comment:
I agree with POIROAE. Let's forget squashing argument.

To me

```rust
trait Migration {
  fn from_version(&self) -> &'static str;
  fn to_version(&self) -> &'static str;
};
```

Describes migration better than:

```rust
trait Migration {
  fn version(&self) -> &'static str;
}
```

Because we can clearly see what version of database this migration expects to find and it's possible to check migration integrity.

```rust
let mut result = BTreeMap::new();
manager.add_migration(Migration0).unwrap();
manager.add_migration(Migration1).unwrap();
manager.add_migration(Migration2).unwrap();
```

and it cannot be achieved without `from` param:

```rust
let mut result = BTreeMap::new();
manager.add_migration(Migration0);
// there is missing migration from 0 to 1, the code still compiles fine
// manager.add_migration(Migration1);
manager.add_migration(Migration2);
```